File: crm-web/src/types/ticket.ts
export type TicketCategory = 'SAV' | 'TECHNIQUE' | 'COMMERCIAL' | 'FACTURATION' | 'LIVRAISON' | 'AUTRE';
export type TicketPriority = 'BASSE' | 'NORMALE' | 'HAUTE' | 'CRITIQUE';
export type TicketStatus   = 'OUVERT' | 'EN_COURS' | 'EN_ATTENTE' | 'RESOLU' | 'FERME';

export interface TicketRelated { id: string; name: string; }
export interface TicketUser    { id: string; fullName: string; email: string; }

export interface TicketComment {
  id: string;
  content: string;
  internal: boolean;
  authorId: string | null;
  authorName: string | null;
  createdAt: string;
}

export interface Ticket {
  id: string;
  reference: string;
  title: string;
  description: string | null;
  category: TicketCategory;
  priority: TicketPriority;
  status: TicketStatus;
  slaDeadline: string | null;
  slaBreached: boolean;
  resolvedAt: string | null;
  closedAt: string | null;
  account: TicketRelated | null;
  contact: TicketRelated | null;
  assignedTo: TicketUser | null;
  createdBy: TicketUser | null;
  comments: TicketComment[];
  createdAt: string;
  updatedAt: string | null;
}

export interface TicketListItem {
  id: string;
  reference: string;
  title: string;
  category: TicketCategory;
  priority: TicketPriority;
  status: TicketStatus;
  slaDeadline: string | null;
  slaBreached: boolean;
  accountId: string | null;
  accountName: string | null;
  contactId: string | null;
  contactName: string | null;
  assignedToId: string | null;
  assignedToName: string | null;
  commentsCount: number;
  createdAt: string;
  resolvedAt: string | null;
}

export interface TicketStats {
  total: number;
  ouverts: number;
  enCours: number;
  enAttente: number;
  resolus: number;
  fermes: number;
  slaBreached: number;
  critiques: number;
  parCategorie: Record<string, number>;
  parPriorite: Record<string, number>;
}

export interface CreateTicketPayload {
  title: string;
  description?: string;
  category?: TicketCategory;
  priority?: TicketPriority;
  accountId?: string;
  contactId?: string;
  assignedToId?: string;
}

export interface UpdateTicketPayload extends CreateTicketPayload {
  status?: TicketStatus;
}

File: crm-backend/crm-core/src/main/java/com/opticrm/core/ticket/entity/Ticket.java
package com.opticrm.core.ticket.entity;

import com.opticrm.core.account.entity.Account;
import com.opticrm.core.contact.entity.Contact;
import com.opticrm.security.entity.User;
import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.UuidGenerator;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@Entity
@Table(name = "tickets")
@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
@EntityListeners(AuditingEntityListener.class)
public class Ticket {

    @Id
    @UuidGenerator
    @Column(name = "id", updatable = false, nullable = false)
    private UUID id;

    @Column(name = "reference", nullable = false, unique = true, length = 20)
    private String reference;

    @Column(name = "title", nullable = false, length = 300)
    private String title;

    @Column(name = "description", columnDefinition = "TEXT")
    private String description;

    @Column(name = "category", length = 50)
    @Enumerated(EnumType.STRING)
    @Builder.Default
    private TicketCategory category = TicketCategory.AUTRE;

    @Column(name = "priority", length = 20)
    @Enumerated(EnumType.STRING)
    @Builder.Default
    private TicketPriority priority = TicketPriority.NORMALE;

    @Column(name = "status", length = 30)
    @Enumerated(EnumType.STRING)
    @Builder.Default
    private TicketStatus status = TicketStatus.OUVERT;

    @Column(name = "sla_deadline")
    private Instant slaDeadline;

    @Column(name = "resolved_at")
    private Instant resolvedAt;

    @Column(name = "closed_at")
    private Instant closedAt;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "account_id")
    private Account account;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "contact_id")
    private Contact contact;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "assigned_to_id")
    private User assignedTo;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "created_by_id")
    private User createdBy;

    @OneToMany(mappedBy = "ticket", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.LAZY)
    @Builder.Default
    private List<TicketComment> comments = new ArrayList<>();

    @CreatedDate
    @Column(name = "created_at", nullable = false, updatable = false)
    private Instant createdAt;

    @LastModifiedDate
    @Column(name = "updated_at")
    private Instant updatedAt;

    // ── Enums ──────────────────────────────────────────────────────────────

    public enum TicketCategory {
        SAV, TECHNIQUE, COMMERCIAL, FACTURATION, LIVRAISON, AUTRE
    }

    public enum TicketPriority {
        BASSE, NORMALE, HAUTE, CRITIQUE
    }

    public enum TicketStatus {
        OUVERT, EN_COURS, EN_ATTENTE, RESOLU, FERME
    }
}
